### DISCUSSION TITLE
Buildx target.matrix CLI override

### DISCUSSION THREAD

**Discussion by Nithos** [docker/buildx#2398] [Q&A] [answered]
Hi, 
I have a bakefile that leverages the matrix to build common images which work great. What I'm wondering about is if there is a way to trigger (or override in the CLI) to build just one of the targets defined in the matrix. I'm hoping to just be able to build one of them during development/debugging rather than all targets defined. The only way I can think of is breaking them apart which negates the benefits of the matrix.

Example
```hcl
target "app" {
  name = "app-${tgt}"
  matrix = {
    tgt = ["foo", "bar"]
  }
  target = tgt
}
``` 
running `docker buildx bake app` will build both `app-foo` and `app-bar` is there a way to only trigger building `foo-bar`?
Looking at the documentation it does not appear that using the `--set` override command will work here (at least not according to the docs).

Any ideas?

**Comment by dvdksn** [comment, marked as answer]
How about using a ternary/variable to check for an override? Something like:

```hcl
variable "MATRIX_OVERRIDE" { default="" }

function "list" {
  params = []
  variadic_params = items
  result = items
}

target "app" {
  name = "app-${tgt}"
  matrix = {
    tgt = notequal("",MATRIX_OVERRIDE) ? list(MATRIX_OVERRIDE) : list("foo", "bar")
  }
  target = tgt
}
```

**Reply by Nithos** [reply]
This is great, yes that should work at least for my current simple case. Would still be nice if there was another way doing this with the current tools (i.e with `--set`).

**Comment by lemeurherve** [comment]
My current workaround is to use the string of a JSON array as environment variable that is then `jsondecode`d in a function to retrieve the tuple for the matrix.

It works but it's kinda ugly:

```hcl
variable "MY_CHOICES" {
  default = "[\"one\",\"two\",\"three\"]"
}

function "mychoices" {
  params = [choices]
  result = jsondecode(choices)
}

target "mytarget" {
  matrix = {
    choice = mychoices(MY_CHOICES)
  }
  name       = "mytarget-${choice}"
}
```
`MY_CHOICES='["four","five"]' docker buildx bake`

I didn't find a way to directly define the tuple content as env var that the HCL parser could resolve as tuple to allow overriding it:

```hcl
variable "mychoices" {
  default = ["one", "two", "three"]
}

target "mytarget" {
  matrix = {
    choice = mychoices
  }
  name       = "mytarget-${choice}"
}
```
Current error:
```console
$ mychoices='["four","five"]' docker buildx bake
 1 | >>> variable "mychoices" {
 2 |       default = ["one", "two", "three"]
 3 |     }
--------------------
ERROR: bake.hcl: Invalid value; unsupported type tuple for variable mychoices
```


I'd love to be able to set `variable`s of type`tuple` or `list` from either env var or either the command line so they could be used in `matrix` without resorting to this trick which is a bit confusing.

**Reply by lemeurherve** [reply]
Ref:
- https://docs.docker.com/build/bake/configuring-build/#from-command-line
- https://docs.docker.com/build/bake/advanced/#typed-variables
- The terraform doc explaining how to declare tuple from env var, letting me hope I could do the same with bake (also using HCL): https://developer.hashicorp.com/terraform/language/values/variables#complex-typed-values

**Reply by lemeurherve** [reply]
I also tried to add a `type` constraint to the variable like we can do with HCL for terraform but the bake HCL parser doesn't recognise it.

https://developer.hashicorp.com/terraform/language/expressions/types
